src/table.js

```diff
@@ -684,10 +758,115 @@ export function __table(source, operations) {
       }))
     );
   }
-  if (primitive) source = source.map((d) => d.value);
   if (source !== input) {
     if (schema) source.schema = schema;
     if (columns) source.columns = columns;
   }
   return source;
 }
+
+function coerceRow(object, types, schema) {
+  const coerced = {};
+  for (const col of schema) {
+    const type = types.get(col.name);
+    const value = object[col.name];
+    coerced[col.name] = type === "raw" ? value : coerceToType(value, type);
+  }
+  return coerced;
+}
+
+function initKey() {
+  return {
+    other: 0,
+    boolean: 0,
+    integer: 0,
+    number: 0,
+    date: 0,
+    string: 0,
+    array: 0,
+    object: 0,
+    bigint: 0,
+    buffer: 0
+  };
+}
+
+// We need to show *all* keys present in the array of Objects
+function getAllKeys(rows) {
+  const keys = new Set();
+  for (const row of rows) {
+    // avoid crash if row is null or undefined
+    if (row) {
+      // only enumerable properties
+      for (const key in row) {
+        // only own properties
+        if (Object.prototype.hasOwnProperty.call(row, key)) {
+          // unique properties, in the order they appear
+          keys.add(key);
+        }
+      }
+    }
+  }
+  return Array.from(keys);
+}
+
+export function inferSchema(source, columns = getAllKeys(source)) {
+  const schema = [];
+  const sampleSize = 100;
+  let sample = source.slice(0, sampleSize);
+  const typeCounts = {};
+  for (const d of sample) {
+    for (const col of columns) {
+      if (!typeCounts[col]) typeCounts[col] = initKey();
+      // for json and sqlite, we already have some types, but for csv and tsv, all
+      // columns are strings here.
+      const type = typeof d[col];
+      const value = type === "string" ? d[col].trim() : d[col];
```

Comment:
Purely a stylistic preference, but consider a form like this:

```js
// Given an arbitrary value, returns "array", "date", "number", "buffer", etc.
function inferType(value) {
```

Then here could say

```js
typeCounts[col][inferType(value)]++;
```

instead of needing separate statements for each type, e.g., instead of

```js
typeCounts[col].array++;
typeCounts[col].date++;
typeCounts[col].buffer++;
// etc.
```

Also, the `inferType` could be exported and tested directly from unit tests, which could make it easier to test.

(I further imagine that `inferType` could return undefined if given a nullish value, or `"null"` or `"undefined"`, or could you check whether the value is nullish before calling `inferType`…)

Reply:
Initially I had some resistance to this, but over the course of trying it out, was won over. Tomorrow I'll add some unit tests for `inferType`. (I think there's value in keeping the unit tests for `inferSchema` that are already in place.) I went with having `inferType` return `"other"` for undefined or nullish values, and I prefer that, but are you thinking of reasons not to do so?

Reply:
> I went with having inferType return "other" for undefined or nullish values, and I prefer that, but are you thinking of reasons not to do so?

Yeah, I don’t think we want to return the `"other"` type for nullish values because it confounds types; `"other"` essentially means an unknown type (some weird type that’s not covered by the other types), and nullish values aren’t weird—they’re common. And pragmatically, to come back to the other comment https://github.com/observablehq/stdlib/pull/346#discussion_r1093733883, if we had 60% nullish values and 40% dates, I expect the inferred type of the column to be `"date"` rather than `"other"`. (I’m commenting without having read the latest version of the code.)

Reply:
That makes sense. I updated to only return "other" when `if (value)` passes.